### PR TITLE
Adding ByteFallback support for `tokenizers`.

### LINKED ISSUE #929: Implement the Byte->char hack of SPM within BPE
The hack has been added here: 

https://github.com/huggingface/tokenizers/pull/896/files#diff-43edb84c28a212d7240c7e278964efe107d223b755acb42e3ab45ac1db2bd26aR377

To conform with the behavior of `SPM`.

Ideally we should find a way to move it out of `BPE` since this is very specific behavior, and we don't want to enable for all moedls (`Unigram`, `WordLevel`, `WordPiece` ) too.

It might prove hard, so the fallback is simply adding it as an option in the BPE config so we can still make the bigscience tokenizers work.

### PR DESCRIPTION
**Pull request by Narsil**
Two items added:

- A flag `byte_fallback` for the `BPE` model. This will be in charge
  of using `<0x61>` instead of unk on unknown tokens.
- A ByteFallback decoder, which will be in charge of putting everything
  back into string whenever possible. Showing � when the byte decoding
  fails (behavior checked against LlamaTokenizer in `transformers`.
  
  
 Fixes #929

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for doing this so quickly!

**Review by McPatate — APPROVED** [review]

File: bindings/python/test.py
[the remark this answers is not in the document]
Comment (Narsil) [reply bindings/python/test.py:1]: Manual leftover . Nice catch !

**Comment by Narsil** [comment]
The bye fallback is working as advertised.

However I still find some differences in tokenization between llama spm and tokenizers after converting one to the other.

It's usually a simple matter of proper configuration, but this particular model is showcasing odd issues, which require much lower level investigation than usual. Hang tight.

**Comment by kjtaed** [comment]
The "byte_fallback" option does not decompose unknown UTF-8 characters into bytes.

Is there example of a training code using byte_fallback?

code

    tokenizer = Tokenizer(BPE(byte_fallback=True))
    tokenizer.pre_tokenizer = pre_tokenizers.Sequence([
        pre_tokenizers.Digits(individual_digits=True),
        pre_tokenizers.Metaspace(),
    ])
    tokenizer.decoder = decoders.Sequence([
        decoders.Metaspace(),
        decoders.ByteFallback(),
    ])
    trainer = trainers.BpeTrainer(
        vocab_size=vocab_size,
        min_frequency=min_frequency,
        special_tokens=special_tokens,
    )
    tokenizer.train(files, trainer=trainer)

**Comment by kjtaed** [comment]
How do I add byte tokens to vocab?

**Comment by Narsil** [comment]
https://huggingface.co/docs/tokenizers/api/trainers

`initial_alphabet=[f'<0x{:02x}>'  for x in range(256)]` for instance ? (You may need to add more depending on the rest of your setup.

**Comment by kjtaed** [comment]
I tried it earlier but byte token is not added.

**the strings contain more than one character, only the first one is kept.**
(https://huggingface.co/docs/tokenizers/api/trainers#tokenizers.trainers.UnigramTrainer.initial_alphabet)

**Comment by Narsil** [comment]
You are linking the UnigramTrainer here, not the BPE trainer.

Which one are you using ? Unigram bytefallback is not yet supported.

**Comment by kjtaed** [comment]
Sorry.

 A list of characters to include in the initial alphabet, even if not seen in the training dataset. If the strings contain more than one character, only the first one is kept.
(https://huggingface.co/docs/tokenizers/api/trainers#tokenizers.trainers.BpeTrainer.initial_alphabet)

Byte tokens cannot be added with 'initial_alphabet'.

If you have example code for 'training BPE with byte_fallback', can you share it?

**Comment by DouglasOrr** [comment]
If it helps, setting `special_tokens=[f"<0x{i:02X}>" for i in range(256)]` [seems to work](https://gist.github.com/DouglasOrr/b5ce3bda0787260051a738739f30b283).

**Comment by chris-ha458** [comment]
setting them as special tokens do not allow them to be trained in certain situations.
This can be hand changed later, but there should be a way to either

1. a way to change a lot of flags for special tokens (changing all 256 of them into special =false)
2. a way to addtokens that wont be 'lost' (if we use the add token method before training, they may or may not be loss)
3. change byte_fallback behavior to deal with b'\x80' like notation since each are technically "one character"
4. when byte_fallback is set, internally add <0x00> ~ <0xFF> tokens into the initial alphabet automatically (what sentencepiece does)

I can think of more, but these seem to be the most effective methods.
